### PR TITLE
refacto(generic-usecase): remove builder, BuilderImpl, DTO and refact…

### PR DESCRIPTION
**Pull request by arnaud-23**
…o request as static creation

generate php 7.4 classes
deletion of RequestBuilder, RequestBuilderImpl and RequestDTO
to use a static create from the request which is no longer an interface but become a class
int that way Request::create()->with()...

### CODE REVIEW

**Comment by yassavic** [comment]
@arnaud-23 Why did you remove Request Builder?
I understand for the DTO.

**Comment by sidux** [comment]
@yassavic I'd think if we don't need interface for the DTO, we probably don't need it either for the builder, (even if for cases when we can have multiple builder implementations for the same DTO that's still very rare)

**Review by sidux — APPROVED** [review]

**Review by romainkuzniak — CHANGES_REQUESTED** [review]
I didn't read all the PR but unsure:
- if you remove the interface, previous DTOs are not suffixed with DTO
- previous DTOs should still be in Request folder in UseCases
- same for builder

**Comment by arnaud-23** [comment]
> I didn't read all the PR but unsure:
> 
> * if you remove the interface, previous DTOs are not suffixed with DTO
> * previous DTOs should still be in Request folder in UseCases
> * same for builder

@romainkuzniak no builder anymore that what we said, to use a static create from the request, isn't it ?

**Comment by arnaud-23** [comment]
@romainkuzniak I don't understand why it should be in UseCase/Request/DTO ?

File: tests/Fixtures/Classes/BusinessRules/Requestors/Domain/SubDomain/GenericUseCaseRequest.php [hunk 1]
```diff
@@ -8,6 +8,10 @@
 
 use OpenClassrooms\UseCase\BusinessRules\Requestors\UseCaseRequest;
 
-interface GenericUseCaseRequest extends UseCaseRequest
+final class GenericUseCaseRequest implements UseCaseRequest
 {
+    public static function create(): self
+    {
+        return new self();
+    }
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/BusinessRules/Requestors/Domain/SubDomain/GenericUseCaseRequest.php:16]: here is the change

File: tests/Fixtures/Classes/BusinessRules/Requestors/Domain/SubDomain/GenericUseCaseRequestBuilder.php [hunk 2]
```diff
@@ -1,16 +0,0 @@
...
-namespace OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Requestors\Domain\SubDomain;
-
-use OpenClassrooms\UseCase\BusinessRules\Requestors\UseCaseRequest;
-
-interface GenericUseCaseRequestBuilder extends UseCaseRequest
-{
-    public function build(): GenericUseCaseRequest;
-
-    public function create(): GenericUseCaseRequestBuilder;
-}
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/BusinessRules/Requestors/Domain/SubDomain/GenericUseCaseRequestBuilder.php:16]: here is the change

File: tests/Fixtures/Classes/BusinessRules/UseCases/Domain/SubDomain/DTO/Request/GenericUseCaseRequestBuilderImpl.php [hunk 3]
```diff
@@ -1,30 +0,0 @@
...
-        return $this->request;
-    }
-
-    public function create(): GenericUseCaseRequestBuilder
-    {
-        $this->request = new GenericUseCaseRequestDTO();
-
-        return $this;
-    }
-}
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/BusinessRules/UseCases/Domain/SubDomain/DTO/Request/GenericUseCaseRequestBuilderImpl.php:30]: here is the change

File: tests/Fixtures/Classes/BusinessRules/UseCases/Domain/SubDomain/DTO/Request/GenericUseCaseRequestDTO.php [hunk 4]
```diff
@@ -1,13 +0,0 @@
...
-
-declare(strict_types=1);
-
-namespace OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\UseCases\Domain\SubDomain\DTO\Request;
-
-use OpenClassrooms\CodeGenerator\Tests\Fixtures\Classes\BusinessRules\Requestors\Domain\SubDomain\GenericUseCaseRequest;
-
-class GenericUseCaseRequestDTO implements GenericUseCaseRequest
-{
-}
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/BusinessRules/UseCases/Domain/SubDomain/DTO/Request/GenericUseCaseRequestDTO.php:13]: here is the change

File: tests/Fixtures/Classes/Tests/BusinessRules/UseCases/Domain/SubDomain/GenericUseCaseTest.php [hunk 5]
```diff
@@ -7,34 +7,21 @@
...
-    /**
-     * @test
-     */
-    final public function genericUseCaseReturnResponse(): void
+    private GenericUseCaseRequest $request;
+
+    private GenericUseCase $useCase;
+
+    /** @test */
+    public function genericUseCaseReturnResponse(): void
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Tests/BusinessRules/UseCases/Domain/SubDomain/GenericUseCaseTest.php:20]: here is the change

File: tests/Fixtures/Classes/Tests/BusinessRules/UseCases/Domain/SubDomain/GenericUseCaseTest.php [hunk 6]
```diff
@@ -43,8 +30,6 @@ protected function setUp(): void
 
     private function buildRequest(): GenericUseCaseRequest
     {
-        return (new GenericUseCaseRequestBuilderImpl())
-            ->create()
-            ->build();
+        return GenericUseCaseRequest::create();
```
Comment (arnaud-23) [inline tests/Fixtures/Classes/Tests/BusinessRules/UseCases/Domain/SubDomain/GenericUseCaseTest.php:33]: here is the change

**Review by yassavic — APPROVED** [review]

**Comment by sidux** [comment]
@arnaud-23 if we respect the CRP principle then we should but the UseCase Request/Response next to files where is it used, and that's the usecases (where we put our DTO's right now) that rule can change if the same request/response is used in multiple usecases from other domains, but that nearly never happens.
So I think we should replace DTO folder by Request and Response folders

**Comment by sidux** [comment]
@romainkuzniak @arnaud-23 I was the one who suggested to use static constructor (create) instead of Builders, my reason was the same as for interfaces that we don't really take advantage of the pattern by sharing the same building sequence between multiple implementations or at least it's really rare.
The advantage of having a static constructor in the other hand is that you'll need to inject only the usecase when using it which makes our dependencies a bit cleaner.

That being said, Im assuming that we all know when we really need a builder to be able to create it when needed.

**Pull request merged**